bibblio.py

```diff
@@ -0,0 +1,391 @@
+
+        # Get any identifiers with uncovered editions in the targeted
+        # CustomList and an associated Work to be recommended.
+        edition_entry = aliased(CustomListEntry)
+        edition_list = aliased(CustomList)
+        qu = qu.join(Work.presentation_edition)\
+                .outerjoin(Work.custom_list_entries)\
+                .outerjoin(CustomListEntry.customlist)\
+                .outerjoin(edition_entry, Edition.custom_list_entries)\
+                .outerjoin(edition_list, edition_entry.customlist)\
+                .filter(or_(
+                    CustomList.id==self.custom_list.id,
+                    edition_list.id==self.custom_list.id))\
+                .options(eagerload(Work.presentation_edition)).distinct()
+
+        if not self.fiction:
+            # Only get nonfiction. This is the default setting.
+            qu = qu.filter(Work.fiction==False)
+
+        return qu
+
+    def process_item(self, work):
+        content_item = self.content_item_from_work(work)
+
+        try:
+            result = self.api.create_content_item(content_item)
+        except Exception as e:
+            return CoverageFailure(
+                work, str(e), data_source=self.output_source,
+                transient=True
+            )
+
+        content_item_id = result.get('contentItemId')
+        bibblio_identifier, _is_new = Identifier.for_foreign_id(
+            self._db, Identifier.BIBBLIO_CONTENT_ITEM_ID, content_item_id
+        )
+
+        work_identifiers = self._db.query(Identifier).filter(
+            Identifier.id.in_(work.all_identifier_ids())
+        )
+        for identifier in work_identifiers:
+            identifier.equivalent_to(
+                self.output_source, bibblio_identifier, 1
+            )
+        return work
+
+    def content_item_from_work(self, work):
+        edition = work.presentation_edition
+
+        name = edition.title + ' by ' + edition.author
+        url = self.edition_permalink(edition)
+        text, data_source = self.get_full_text(work)
+        provider = dict(name=data_source.name)
+
+        content_item = dict(
+            name=name, url=url, text=text, provider=provider
+        )
+        if self.catalogue_id:
+            content_item['catalogueId'] = self.catalogue_id
+
+        return content_item
+
+    @classmethod
+    def edition_permalink(cls, edition):
+        """Gets a unique URL for the target Work"""
+
+        base_url = Configuration.integration_url(
+            Configuration.CONTENT_SERVER_INTEGRATION, required=True
+        )
+        scheme, host = urlparse(base_url)[0:2]
+        base_url = '://'.join([scheme, host])
+
+        urn = edition.primary_identifier.urn
+        initialization_value = os.environ.get('AUTOINITIALIZE')
+        try:
+            os.environ['AUTOINITIALIZE'] = 'False'
+            from app import app
+            with app.test_request_context(base_url=base_url):
```

Comment:
It's not great but we can make it easier to test this sort of thing in the future.